### DISCUSSION TITLE
Request: Add the "Background Blur" feature from Nvidia Broadcast to StreamFX

### DISCUSSION THREAD

**Discussion by noiceFTW** [Xaymar/obs-stream-effects#854] [Ideas & Feedback]
### Explain the Feature/Enhancement

The background blur feature from nvidia broadcast allows you to blur the background while keeping your subject focused.

### What makes this feature/enhancement useful for StreamFX and its users?

This is quite convenient as it removes the need to use Nvidia Broadcast and bring Broadcast into OBS as a video source. The current Broadcast software also only supports a few resolutions, which are most likely different from what users typically use, as they vary greatly. This is also beneficial as it would complement other Nvidia Broadcast features which already exist such as auto framing and virtual greenscreen. 

### Checklist

- [ ] I am offering a monetary bounty/reward for the implementation of this Feature/Enhancement.
- [ ] There is no other way to currently do this in StreamFX or OBS Studio.

**Comment by Xaymar** [comment]
Technically this is already possible entirely with StreamFX and OBS Studio, just not "user friendly".

**Edit:** Since I'm sure people will ask how:

1. Add the original video input.
2. Use Source Mirror to create two clones, we'll call them A and B for now.
3. Add Virtual Greenscreen to A, configure according to your needs.
4. Add Render Delay set to exactly 2 frames to B.
5. Add Dynamic Mask to B and point it at A, set it to create a negative mask (if A is visible, B is not).
6. Add a Shader to B that fills up the now empty area with the nearest surrounding color.
7. Add a Blur (Dual Filtering) to B
8. ???
9. Profit

**Comment by spectangulation** [comment]
> Technically this is already possible entirely with StreamFX and OBS Studio, just not "user friendly".
> 
> **Edit:** Since I'm sure people will ask how:
> 
> 1. Add the original video input.
> 2. Use Source Mirror to create two clones, we'll call them A and B for now.
> 3. Add Virtual Greenscreen to A, configure according to your needs.
> 4. Add Render Delay set to exactly 2 frames to B.
> 5. Add Dynamic Mask to B and point it at A, set it to create a negative mask (if A is visible, B is not).
> 6. Add a Shader to B that fills up the now empty area with the nearest surrounding color.
> 7. Add a Blur (Dual Filtering) to B
> 8. ???
> 9. Profit

it's easy than this:
1. add camera
2. add source mirror of camera
3. arrange source mirror on top of camera
4. set v-greenscreen on source mirror
5. set blur streamFX to camera

done.

just tranform them so their centers always align and put them in a group so they never move from one another :)

**Comment by HunterAP23** [comment]
> it's easy than this:

add camera
add source mirror of camera
arrange source mirror on top of camera
set v-greenscreen on source mirror
set blur streamFX to camera
done.

just tranform them so their centers always align and put them in a group so they never move from one another :)

You can skip steps 5 and 6 in Xaymar's instructions, and even step 4, but that one actually helps get rid of some ghosting-like effects that happen between you and the blurred background

**Comment by Xaymar** [comment]
> just tranform them so their centers always align and put them in a group so they never move from one another :)

1. Render Delay is required to synchronize the blurred copy with the masked copy. Otherwise you will get some significant difference in what is blurred and what is masked.
2. The "replace empty area with nearby pixels" is required in order to correctly mimic the blurred background mode of the Tensor effect.

You can always skip things, just don't come complaining if it doesn't look the same.

**Comment by LuminairPrime** [comment]
Thank you for the suggestion!

> set it to create a negative mask

I don't know how to do this step, or what you mean by negative. The Dynamic Mask filter has a lot of bars to change :P And when I change the settings back and forth to brute force the solution, things get mixed up with the filter. Manually dragging the bars back to default doesn't work and I need to delete the mask.

**Comment by spectangulation** [comment]
> You can always skip things, just don't come complaining if it doesn't look the same.

i dont do these things, and i don't witness any issues... can u provide something that shows, visually, the differences side by side?

**Comment by LuminairPrime** [comment]
> can u provide something that shows, visually, the differences side by side?

Don't waste his time please ;) If it's working for you, then that's enough talking about it. It's more important to work on the project than debate these things.

**Comment by testingthisout12345** [comment]
> Technically this is already possible entirely with StreamFX and OBS Studio, just not "user friendly".
> 
> **Edit:** Since I'm sure people will ask how:
> 
> 1. Add the original video input.
> 2. Use Source Mirror to create two clones, we'll call them A and B for now.
> 3. Add Virtual Greenscreen to A, configure according to your needs.
> 4. Add Render Delay set to exactly 2 frames to B.
> 5. Add Dynamic Mask to B and point it at A, set it to create a negative mask (if A is visible, B is not).
> 6. Add a Shader to B that fills up the now empty area with the nearest surrounding color.
> 7. Add a Blur (Dual Filtering) to B
> 8. ???
> 9. Profit

I know this is an old thread, but how do I do the negative mask & which shader is that?

**Comment by LuminairPrime** [comment]
@TheJAXi To answer your question, I have no idea LUL, but I'm going to try following these instructions again to figure it out. 

@Xaymar If you want to help, we're not sure if there's a bug or we're just stupid in step 5: negative mask 👍 

I'm using the latest alpha build in case that'll help me. I'm taking a guess that the negative mask step means to drag the four channel base values down to -100. Now I don't know what shader to add for step 6, lol! ;)

**Comment by testingthisout12345** [comment]
> @TheJAXi To answer your question, I have no idea LUL, but I'm going to try following these instructions again to figure it out.
> 
> @Xaymar If you want to help, we're not sure if there's a bug or we're just stupid in step 5: negative mask 👍
> 
> I'm using the latest alpha build in case that'll help me. I'm taking a guess that the negative mask step means to drag the four channel base values down to -100. Now I don't know what shader to add for step 6, lol! ;)


Aight? Can you msg here if you figure it out. It'd be cool if they added the nvidia filter as it's easy & "good" looking.

I kinda don't need it with my DSLR, but sometimes I like to blur everything in the background (got no greenscreen & I think blurring is better for me as it creates depth? when you still can see background shapes, green screens just make everything look flat).

**Comment by Xaymar** [comment]
Please stop derailing the issue
